Support multiple multiple GitHub applications on the same installation

chmouel opened the pull request:
This introduce the concept of multiple GitHub application from the same PaC installations.

You will need to deploy another controller with another set of environement variable pointing to a new secret and a configmap to configure this, see the documentation showing how to install this.

e2e tests now tests concurrently (for some tests) on GHE instance (https://ghe.pipelinesascode.com/) and public GitHub.

The watcher is unique it read the information from where the controller/secret/configmap the request come from on the PipelineRun annotation and resolves it to the right place with the right secret to the right place.

We now stores as well some of of the controller info in the context instead of the global variables, this involves a large refactoring around the code.


Tasks:

- [x] Docs documenting installation
- [x] Script to deploy a new controller easily for local and testing
- [x]  Run E2E on a https://ghe.pipelinesascode.com/ app and test concurrently if watcher do the right thing
- [X] E2E
- [X] Unit
- [X] passing by context 
- [X] Stress testing concurrent access https://www.youtube.com/watch?v=E977dyRwm6w&t=1s

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 A good commit message is important for other reviewers to understand the context of your change. Please refer to [How to Write a Git Commit Message](https://cbea.ms/git-commit/) for more details how to write beautiful commit messages. We rather have the commit message in the PR body and the commit message instead of an external website.
- [ ] ♽ Run `make test` before submitting a PR (ie: with [pre-commit](https://pipelinesascode.com/dev/tools), no need to waste CPU cycle on CI. (or even better install [pre-commit](https://pre-commit.com/) and do `pre-commit install` in the root of this repo).
- [ ] ✨ We heavily rely on linters to get our code clean and consistent, please ensure that you have run `make lint` before submitting a PR. The [markdownlint](https://github.com/DavidAnson/markdownlint) error can get usually fixed by running `make fix-markdownlint` (make sure it's installed first)
- [ ] 📖 If you are adding a user facing feature or make a change of the behavior, please verify that you have documented it
- [ ] 🧪 100% coverage is not a target but most of the time we would rather have a unit test if you make a code change.
- [ ] 🎁 If that's something that is possible to do please ensure to check if we can add a e2e test.
- [ ] 🔎 If there is a flakiness in the CI tests then don't *necessary* ignore it, better get the flakyness fixed before merging or if that's not possible there is a good reason to bypass it. (token rate limitation may be a good reason to skip).

piyush-garg committed:
Fix flaky unit tests

This fixes the flaky unit tests, as they were failing
because of the order of the elements of array

chmouel committed:
Multi namespace support

This introduce the concept of multiple GitHub application from the same
PaC installations.

You will need to deploy another controller with another set of
environment variable pointing to a new secret and a ConfigMap to
configure this, see the documentation showing how to install this.

e2e tests now tests concurrently (for some tests) on GHE instance
(ghe.pipelinesascode.com) and public GitHub.

The watcher is unique it read the information from where the
controller/secret/configmap the request come from on the PipelineRun
annotation and resolves it to the right place with the right secret to
the right place.

We now stores as well some of of the controller info in the context
instead of the global variables, this involves a large refactoring
around the code.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

add openshift routes for second-controller.py

Add support for output openshift routes for second-controller.py

chmouel commented:
/retest

chmouel commented:
I haven't squashed the commit yet, will do before merge, since it's easier to manage like this i think.. let me know if it doesn't for the reviewers

sm43 commented on hack/second-controller.py:
```diff
@@ -0,0 +1,186 @@
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+    parser.add_argument(
+        "label",
```
```suggestion
        "--label",
``` 
to be consistent with other?

sm43 replied:
ah I see its not required 👍

chmouel replied:
yeah :D dashes `--` are for non required stuff and just a arg make it required, and since we need a label for all things then it's an argument

sm43 commented on pkg/adapter/adapter.go:
```diff
@@ -101,15 +100,15 @@ func (l *listener) Start(ctx context.Context) error {
-		// Force WatchConfigMapChanges go routines to actually start
-		<-c
+		// we should fix this, this basically reads configmap on every request, is that supposed to be okay?
+		currentControllerName := info.GetCurrentControlerName(ctx)
```
```suggestion
		currentControllerName := info.GetCurrentControllerName(ctx)
```

sm43 commented on pkg/adapter/adapter.go:
```diff
@@ -101,15 +100,15 @@ func (l *listener) Start(ctx context.Context) error {
-		}()
-		// Force WatchConfigMapChanges go routines to actually start
-		<-c
+		// we should fix this, this basically reads configmap on every request, is that supposed to be okay?
```
i think switching to viper will fix this

chmouel replied:
yes that's the idea :)

sm43 commented on pkg/reconciler/controller.go:
```diff
@@ -19,33 +20,34 @@ import (
-			}
-		}()
-		<-c
+		if err := run.UpdatePACInfo(ctx); err != nil {
```
with this any new changes to configmap, reconciler will not pick up them

chmouel replied:
yeah... I don't mind loosing that functionality until we get viper in....

sm43 replied:
okay just in case someone reports a bug ask them restart the pod 😄

chmouel replied:
yeah that's the idea 🙃 

but i think hopefully we can get viper merged before soon, so that not impact people too much

chmouel replied:
I changed it to keep the old behaviour (reading the configmap on every request/reconcilation)...

sm43 reviewed (approved):
✅

chmouel committed:
Parse configmap on every reconcilation

Keep the previous behaviour as before, it's bad for performance but
hopefully we can fix that with another pr

chmouel commented:
Squashed all commit, and kept the old behaviour with regard to the reading of the configmap every time

savitaashture commented:
Hi @chmouel 

I am able to test the PR changes manually by deploying second controller using `python3 ./hack/second-controller.py LABEL|kubectl -f-`

I have a questions:

1. In order to create GitHubApp for second controller user cannot make use of `tkn pac bootstrap` instead need to setup manually so question is can we update tkn pac bootstrap to setup for second GithubApp as well :thinking: 
2. Similarly for `tkn pac create repo`

chmouel commented:
@savitaashture CLI support is not in scope for this PR, this greatly amplify the complexity the cli implementation (which is already complex) and the primary user asking for this feature does not need cli integration...

ultimately it's the operator role to be able to handle this....

savitaashture commented:
> @savitaashture CLI support is not in scope for this PR, this greatly amplify the complexity the cli implementation (which is already complex) and the primary user asking for this feature does not need cli integration...
> 
> ultimately it's the operator role to be able to handle this....

Ah i see

Can we document this ? so that there will be clear indication

savitaashture commented on hack/second-controller.py:
```diff
@@ -0,0 +1,185 @@
+
+It's designed to be used like this:
+
+kubectl apply -f <(python3 yamlconvert.py LABEL)
```
as its a help message do we want to provide proper name so that user know what needs to be give there 
```suggestion
kubectl apply -f <(python3 second-controller.py LABEL)
```

chmouel replied:
what do you mean by proper name? your help fix, right?

savitaashture commented on hack/second-controller.py:
```diff
@@ -0,0 +1,185 @@
+variable set properly) use ko to deploy the second controller image in the
+target namespace (default: pipelines-as-code):
+
+ko apply -f <(python3 yamlconvert.py --controller-image=ko LABEL)
```
```suggestion
ko apply -f <(python3 second-controller.py --controller-image=ko LABEL)
```

savitaashture commented:
The default generated image after running this script 
```
python3 ./hack/second-controller.py LABEL
```
is `ghcr.io/openshift-pipelines/pipelines-as-code-controller:stable` 

when i apply second controller yaml and raise a Pull Request getting below error
```
{"level":"error","ts":"2023-12-20T10:41:40.309Z","logger":"pipelinesascode","caller":"adapter/sinker.go:29","msg":"failed to parse event: could not refresh installation id 3's token: received non 2xx response status \"404 Not Found\" when fetching https://ghe.pipelinesascode.com/api/v3/app/installations/3/access_tokens","commit":"7215808","provider":"github","event-id":"5b962670-9f24-11ee-974f-de4f79de9253","stacktrace":"github.com/openshift-pipelines/pipelines-as-code/pkg/adapter.(*sinker).processEventPayload\n\t/src/pkg/adapter/sinker.go:29\ngithub.com/openshift-pipelines/pipelines-as-code/pkg/adapter.(*sinker).processEvent\n\t/src/pkg/adapter/sinker.go:46\ngithub.com/openshift-pipelines/pipelines-as-code/pkg/adapter.listener.handleEvent.func1.2\n\t/src/pkg/adapter/adapter.go:190"}
{"level":"error","ts":"2023-12-20T10:41:40.309Z","logger":"pipelinesascode","caller":"adapter/adapter.go:192","msg":"an error occurred: could not refresh installation id 3's token: received non 2xx response status \"404 Not Found\" when fetching https://ghe.pipelinesascode.com/api/v3/app/installations/3/access_tokens","commit":"7215808","provider":"github","event-id":"5b962670-9f24-11ee-974f-de4f79de9253","stacktrace":"github.com/openshift-pipelines/pipelines-as-code/pkg/adapter.listener.handleEvent.func1.2\n\t/src/pkg/adapter/adapter.go:192"}

```

But when i replace with image built by `ko` its working fine 

So question is does that image is working for you ?

chmouel commented:
>  So question is does that image is working for you ?

when this PR will be merged, the image will be generated and it will work on the stable image.....

savitaashture commented on hack/second-controller.py:
```diff
@@ -0,0 +1,185 @@
+                if env["name"] == "PAC_CONTROLLER_CONFIGMAP":
+                    env["value"] = args.configmap
+
+service = {}
+with open("config/401-controller-service.yaml", "r", encoding="utf-8") as f:
+    service = yaml.load(f, Loader=yaml.FullLoader)
+    service["spec"]["selector"]["app.kubernetes.io/name"] = args.label + "-controller"
+    service["metadata"]["name"] = args.label + "-controller"
+    service["metadata"]["labels"]["app"] = args.label
+    service["metadata"]["labels"]["app.kubernetes.io/name"] = args.label + "-controller"
+
```
It would be beneficial to also incorporate this for Openshift Route, I mean mirroring the approach used for the service.

i.e
```
route = {}
with open("config/openshift/10-routes.yaml", "r", encoding="utf-8") as f:
```

chmouel replied:
that would makes things specific to openshift then?

savitaashture replied:
That's true 

How about determining, based on the provided parameter, whether to generate a route or not? However, this is more of an optional enhancement to make it easy for OpenShift. If you believe it might add complexity with user input, we can skip implementing it."

chmouel replied:
i added and option in latest commit,

chmouel commented:
/retest

piyush-garg commented:
> Hi @chmouel
> 
> I am able to test the PR changes manually by deploying second controller using `python3 ./hack/second-controller.py LABEL|kubectl -f-`
> 
> I have a questions:
> 
> 1. In order to create GitHubApp for second controller user cannot make use of `tkn pac bootstrap` instead need to setup manually so question is can we update tkn pac bootstrap to setup for second GithubApp as well 🤔
> 2. Similarly for `tkn pac create repo`

@savitaashture I created the github app manually, but `tkn pac create repo ` command works fine

piyush-garg commented:
lgtm

chmouel commented:
/retest

chmouel committed:
Only log err when failing to rm gitea test repo

We are deleting the repo at the end of the test, if it fails to delete
it is not a big deal, it just for hygienic purpose, so we should not
fail

chmouel commented:
/retest

random unittests failure which occur more and more fileed issue in #1547 

(appeared in other PR too)

savitaashture commented:
> tkn



> > ultimately it's the operator role to be able to handle this....
> 
> Ah i see
> 
> Can we document this ? so that there will be clear indication

looks good to me 

https://github.com/openshift-pipelines/pipelines-as-code/pull/1518#issuecomment-1864254997
wdyt of documenting this
i mean for second controller GithubApp setup has to be done manual only

chmouel commented:
@savitaashture it's already document in the help string, and the script is further documented in the docs package, is there anything i have missed ?

savitaashture commented:
> @savitaashture it's already document in the help string, and the script is further documented in the docs package, is there anything i have missed ?

help string documents how to use python script

But my point is we need to mention that for second controller creating Github App is manual and there won't be any support from CLI side
